qiskit/pulse/commands/parametric_pulses.py

```diff
@@ -0,0 +1,337 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Parametric pulse commands. These are pulse commands which are described by a known formula and some
+parameters.
+
+If a backend supports parametric pulses, it will have the attribute
+`backend.configuration().parametric_pulses`, which is a list of supported pulse shapes, such as
+`['gaussian', 'gaussian_square', 'drag']`. A Pulse Schedule, using parametric pulses, which is
+assembled for a backend which supports those pulses, will result in a Qobj which is dramatically
+smaller than one which uses SamplePulses.
+
+This module can easily be extended to describe more pulse shapes. The new class should:
+  - have a descriptive name
+  - be a well known and/or well described formula (include the formula in the class docstring)
+  - take some parameters (at least `duration`) and validate them, if necessary
+  - implement a `get_sample_pulse` method which returns a corresponding SamplePulse in the
+    case that it is assembled for a backend which does not support it.
+
+The new pulse must then be registered by the assembler in
+`qiskit/qobj/converters/pulse_instruction.py:ParametricPulseShapes`
+by following the existing pattern:
+    class ParametricPulseShapes(Enum):
+        gaussian = commands.Gaussian
+        ...
+        new_supported_pulse_name = commands.YourPulseCommandClass
+"""
+from abc import abstractmethod
+from typing import Any, Dict, Optional
+import math
+
+import numpy as np
+
+from qiskit.pulse.channels import PulseChannel
+from qiskit.pulse.exceptions import PulseError
+from qiskit.pulse.pulse_lib.discrete import gaussian, gaussian_square, drag, constant
+from qiskit.pulse.pulse_lib import continuous
+
+from .command import Command
+from .sample_pulse import SamplePulse
+from .instruction import Instruction
+
+# pylint: disable=missing-docstring
+
+
+class ParametricPulse(Command):
+    """The abstract superclass for parametric pulses."""
+    prefix = 'p'
+
+    @abstractmethod
+    def __init__(self, duration: int):
```

Comment:
I think we had discussed a few ways to do this. I don't like `**args` because I want each required parameter to be clearly documented. Could set `self.parameters` inside `__init__`, it's just long and doesn't carry meaning. Seems like something that should be fillable programmatically, but I could understand that you'd worry about it being robust as is. It's annoying because it is only necessary for assembling, and I'd rather not need it at all in this file. Do you have other ideas?

Reply:
Moving this comment into this one thread: "I think it would be better if there was no magic happening here as this could get confusing for a more complicated pulse and private variables."

Reply:
I think it would be better to make it explicit over implicit. So I would prefer it being provided by the programmer. If it was in `**kwargs**` the programmer can still document it in their class constructor.

Reply:
I prefer explicit, too, and letting python check that the required args are there. I like the signature `func(a: type, b: type)` over `func(*args)` so we don't have to check for `a` and `b`

Reply:
Yes, I was saying the Abstract class could support that but change the implementing class to be specific. This violates LSP though.

Reply:
So an abstract `parameters` property might be better.

Reply:
I feel like I misunderstood this entire thread. You just want this line to say `*parameters` rather than `duration`, and you don't want `duration` to show? I think I was influenced at the time by this `    def __init__(self, duration: Union[int, np.integer] = None):` in the `Command` class

Reply:
So, this isn't about all the init functions, just the abstract one?

Reply:
It's not really about this line per-say. It is about the way `parameters` are made available to users and the automated `get_params` method. I think an abstract property `parameters` is what I am advocating for.

Reply:
Okay, I agree with you. Updating 👍